controllers/machinedeployment_ondelete.go

```diff
@@ -0,0 +1,146 @@
+		return err
+	}
+
+	if mdutil.DeploymentComplete(d, &d.Status) {
+		if err := r.cleanupDeployment(ctx, oldMSs, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//reconcileOldMachineSetsOnDelete handles reconciliation of Old MachineSets associated with the MachineDeployment in the OnDelete MachineDeploymentStrategyType
+func (r *MachineDeploymentReconciler) reconcileOldMachineSetsOnDelete(ctx context.Context, oldMSs []*clusterv1.MachineSet, allMSs []*clusterv1.MachineSet, deployment *clusterv1.MachineDeployment) error {
+	log := ctrl.LoggerFrom(ctx)
+	if deployment.Spec.Replicas == nil {
+		return errors.Errorf("spec replicas for MachineDeployment %q is nil, this is unexpected", client.ObjectKeyFromObject(deployment).String())
+	}
+	log.V(4).Info("Checking to see if machines have been deleted or are in the process of deleting for old machine sets")
+	totalReplicas := mdutil.GetReplicaCountForMachineSets(allMSs)
+	scaleDownAmount := totalReplicas - *deployment.Spec.Replicas
+	for _, oldMS := range oldMSs {
+		if oldMS.Spec.Replicas == nil || *oldMS.Spec.Replicas <= 0 {
+			log.V(4).Info(fmt.Sprintf("MachineSet %s fully scaled down", oldMS.Name))
+			continue
+		}
+		selectorMap, err := metav1.LabelSelectorAsMap(&oldMS.Spec.Selector)
+		if err != nil {
+			log.V(4).Error(err, "failed to convert MachineSet %q label selector to a map", oldMS.Name)
+			continue
+		}
+		// Get all Machines linked to this MachineSet.
+		allMachinesInOldMS := &clusterv1.MachineList{}
+		err = r.Client.List(ctx,
+			allMachinesInOldMS,
+			client.InNamespace(oldMS.Namespace),
+			client.MatchingLabels(selectorMap),
+		)
+		if err != nil {
+			return errors.Wrap(err, "failed to list machines")
+		}
+		updatedReplicaCount := oldMS.Status.Replicas - mdutil.GetDeletingMachineCount(allMachinesInOldMS)
+		if updatedReplicaCount < 0 {
+			return errors.Errorf("negative updated replica count %d for MachineSet %q, this is unexpected", updatedReplicaCount, client.ObjectKeyFromObject(oldMS).String())
+		}
+		machineSetScaleDownAmountDueToMachineDeletion := *oldMS.Spec.Replicas - updatedReplicaCount
+		if machineSetScaleDownAmountDueToMachineDeletion < 0 {
+			log.V(4).Error(errors.Errorf("unexpected negative scale down amount: %d", machineSetScaleDownAmountDueToMachineDeletion), fmt.Sprintf("Error reconciling MachineSet %s", oldMS.Name))
+		}
+		scaleDownAmount -= machineSetScaleDownAmountDueToMachineDeletion
+		log.V(4).Info(fmt.Sprintf("Scaling MachineSet %s to %d", oldMS.Name, updatedReplicaCount))
+		if err := r.scaleMachineSet(ctx, oldMS, updatedReplicaCount, deployment); err != nil {
+			return err
+		}
+	}
+	log.V(4).Info("Finished reconcile of Old MachineSets to account for deleted machines. Now analyzing if there's more potential to scale down")
+	for _, oldMS := range oldMSs {
+		if scaleDownAmount <= 0 {
+			break
+		}
+		if oldMS.Spec.Replicas == nil || *oldMS.Spec.Replicas <= 0 {
+			log.V(4).Info(fmt.Sprintf("MachineSet %s fully scaled down", oldMS.Name))
```

Comment:
nit: For all sprintf logging we should preferably follow structured logging conventions and use key/value format https://github.com/kubernetes-sigs/controller-runtime/blob/master/TMP-LOGGING.md
e.g
`log.V(4).Info("fully scaled down", "machineSet", ctrlclient.ObjectKeyFromObject(oldMS).String())))`

Reply:
these changes caused panics during api testing. Will not be able to port to new format and will need to keep consistent with old

Reply:
addressing

Reply:
addressed